pkg/reconciliation/construct_service_test.go

```diff
@@ -407,3 +408,99 @@ func TestAddingAdditionalLabels(t *testing.T) {
 		t.Errorf("service labels = %v, want %v", service.Labels, expected)
 	}
 }
+
+func TestServicePorts(t *testing.T) {
+	tests := []struct {
+		name                string
+		dc                  *api.CassandraDatacenter
+		dcServicePorts      []int32
+		allPodsServicePorts []int32
+	}{
+		{
+			name: "Cassandra 3.11.14",
+			dc: &api.CassandraDatacenter{
+				Spec: api.CassandraDatacenterSpec{
+					ClusterName:   "bob",
+					ServerType:    "cassandra",
+					ServerVersion: "3.11.14",
+				},
+			},
+			dcServicePorts:      []int32{8080, 9000, 9042, 9103, 9142, 9160},
+			allPodsServicePorts: []int32{8080, 9000, 9042, 9103},
+		},
+		{
+			name: "Cassandra 4.0.7",
+			dc: &api.CassandraDatacenter{
+				Spec: api.CassandraDatacenterSpec{
+					ClusterName:   "bob",
+					ServerType:    "cassandra",
+					ServerVersion: "4.0.7",
+				},
+			},
+			dcServicePorts:      []int32{8080, 9000, 9042, 9103, 9142},
+			allPodsServicePorts: []int32{8080, 9000, 9042, 9103},
+		},
+		{
+			name: "DSE 6.8.31",
+			dc: &api.CassandraDatacenter{
+				Spec: api.CassandraDatacenterSpec{
+					ClusterName:   "bob",
+					ServerType:    "dse",
+					ServerVersion: "6.8.31",
+				},
+			},
+			dcServicePorts:      []int32{8080, 9000, 9042, 9103, 9142, 9160},
+			allPodsServicePorts: []int32{8080, 9000, 9042, 9103},
+		},
+		{
+			name: "Cassandra 4.0.7 with custom ports",
+			dc: &api.CassandraDatacenter{
+				Spec: api.CassandraDatacenterSpec{
+					ClusterName:   "bob",
+					ServerType:    "cassandra",
+					ServerVersion: "4.0.7",
+					PodTemplateSpec: &corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Name: "cassandra",
+									Ports: []corev1.ContainerPort{
+										{
+											Name:          "metrics",
+											ContainerPort: 9004,
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			// FIXME: 9004 should be in the list of open ports
```

Comment:
This is interesting: I copied this test from `TestPorts`, but I realized that if the user changes a managed port at container level, the change is not reflected at service level.